rfcs/20200731-Using-Arm-Compute-Library-kernels-on-AArch64/README.md

```diff
@@ -0,0 +1,315 @@
+...
+acl_data_.acl_src.allocator()->import_memory(const_cast<data_t *>(src_base));
+~~~
+
+
+## Limitations
+
+The implementation outlined in this RFC has a number of limitations which could
+be addressed in future PRs:
+
+- Only the forward propagation regime is supported (dnnl\_forward\_training,
+dnnl\_forward\_inference, dnnl\_forward\_scoring);
+- Only Conv2D is supported. There is no support for Conv3D, Conv1D or depthwise
+convolutions;
+- Grouped convolutions are currently not supported in ArmCL;
+- Only fp32 datatype is supported at the moment;
+- ArmCL requires all the tensors to be in the same data layout, NCHW or NHWC.
+Blocked memory formats are not supported;
+- Only one eltwise operation is supported, activation
+
+## Results/validation
+oneDNN's standard test suites were used for validation. Some of the tests there
+compare the output with a reference variant like:
+
+~~~c++
+compare_data<data_t_dst>(dst_ref, c_dst, 1e-2);
+~~~
+
+and fail if it is incorrect (for example,
+[test\_convolution\_eltwise\_forward](https://github.com/oneapi-src/oneDNN/blob/master/tests/gtests/test_convolution_eltwise_forward_common.hpp#L215-L221),
+[test\_deconvolution](https://github.com/oneapi-src/oneDNN/blob/master/tests/gtests/test_deconvolution.cpp#L423-L429),
+[test\_convolution\_forward\_f32](https://github.com/oneapi-src/oneDNN/blob/master/tests/gtests/test_convolution_forward_common.hpp#L206-L212) etc.). ArmCL-base convolution is
+called in a number of tests with output validation, covering cases with
+strides > 1,
+dilations, padding ([test\_global\_scratchpad](https://github.com/oneapi-src/oneDNN/blob/master/tests/gtests/test_global_scratchpad.cpp#L93-L94)
+also checks the case with asymmetric padding), different activation functions.
+It passes these tests, meaning than it
+provides the same functionality as `CPU_INSTANCE(gemm_convolution_fwd_t)` for
+the same launch parameters:
+
+```
+./tests/gtests/test_convolution_forward_f32
+./tests/gtests/test_convolution_eltwise_forward_f32
+./tests/gtests/test_deconvolution
+./tests/gtests/test_global_scratchpad
+```
+
+Overall, all 98 oneDNN tests passed, the testing was performed on AArch64 CPU with `ctest`
+and the build
+
+```
+cmake [...] -DCMAKE_BUILD_TYPE=Release -DDNNL_AARCH64_USE_ARMCL=1
+```
+
+
+## Expected performance benefits
+
+This proposal is intended to lay the groundwork to enable the development of
+more performant AArch64 implementation which leverages the ArmCL library. We
+expect the ongoing work to optimize ArmCL's implementations on infrastructure
+scale CPUs and high core counts may expose opportunities to adapt ArmCL's API in
+the longer term. Preliminary tests demonstrated noticeable increase of
+performance in single-threaded regime on Thunder X2 AArch64 CPU.
+
+## Limited impact
+
+The scope of this RFC, and the PRs outlined below, is limited such that:
+
+- There are no changes to the API;
+- There is no impact on non-AArch64 builds.
+
+## Implementation plan
```

Comment:
How new configurations will be handled in CI?

Reply:
We have updated our downstream CI to include builds on AArch64 both with and without the ArmCL dependency. At this stage, we intend to ensure that the 'default' AArch64 build is unaffected - i.e. that oneDNN will build without ArmCL "out of the box", with the ArmCL dependence (which required an env var to provide the lib location) requiring a build-time option

The upstream Drone CI will require a similar update - we'll note this on the Implementation Plan in the RFC.

Should the update to Drone CI be included as a commit in a PR to implement these changes, or as a separate PR to follow?

Reply:
Thanks, @nSircombe. Including the CI configuration changes into initial PR or separate PR is up to you.